% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/percentVJ.R
\name{percentVJ}
\alias{percentVJ}
\title{Examining the V or J gene usage across groupings}
\usage{
percentVJ(
  df,
  chain = "TRB",
  group.by = NULL,
  split.by = NULL,
  exportTable = FALSE,
  palette = "inferno",
  pcoa.group.by = NULL,
  pcoa = FALSE,
  dist.method = "manhattan",
  point.size = 3
)
}
\arguments{
\item{df}{The product of \code{\link{combineTCR}}, \code{\link{combineBCR}}, or
\code{\link{combineExpression}}.}

\item{chain}{"TRA", "TRB", "TRG", "TRG", "IGH", "IGL"}

\item{group.by}{The column header used for grouping.}

\item{split.by}{If using a single-cell object, the column header 
to group the new list. NULL will return clusters.}

\item{exportTable}{Returns the data frame used for forming the graph}

\item{palette}{Colors to use in visualization - input any hcl.pals()}
}
\value{
ggplot of percentage of V and J gene pairings as a heatmap
}
\description{
This function the proportion V or J genes used by 
grouping variables
}
\examples{
#Making combined contig data
combined <- combineTCR(contig_list, 
                        samples = c("P17B", "P17L", "P18B", "P18L", 
                                    "P19B","P19L", "P20B", "P20L"))
percentVJ(combined, chain = "TRB")

}
