% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/vizGenes.R
\name{vizGenes}
\alias{vizGenes}
\title{Visualizing the distribution of gene usage}
\usage{
vizGenes(
  df,
  gene = "V",
  chain = "TRA",
  plot = "heatmap",
  y.axis = "sample",
  order = "gene",
  scale = TRUE,
  group.by = NULL,
  pcoa.group.by = NULL,
  split.by = NULL,
  exportTable = FALSE,
  palette = "inferno",
  dist.method = "manhattan",
  point.size = 3
)
}
\arguments{
\item{df}{The product of \code{\link{combineTCR}}, \code{\link{combineBCR}}, or
\code{\link{combineExpression}}.}

\item{gene}{Which part of the immune receptor to visualize - V, D, J, C}

\item{chain}{indicate the specific chain should be used - 
e.g. "TRA", "TRG", "IGH", "IGL" (no both option here)}

\item{plot}{The type of plot to return - heatmap, bar or pcoa}

\item{y.axis}{Variable to separate the y-axis, can be both categorical or other gene 
gene segments such as V, D, J, or C.}

\item{order}{Categorical variable to organize the x-axis, either "gene" or "variance"}

\item{scale}{Converts the individual count of genes to proportion using the total 
respective repertoire size}

\item{group.by}{The column header used for grouping.}

\item{pcoa.group.by}{The column header used for grouping in PCoA.}

\item{split.by}{If using a single-cell object, the column header 
to group the new list. NULL will return clusters.}

\item{exportTable}{Returns the data frame used for forming the graph.}

\item{palette}{Colors to use in visualization - input any hcl.pals()}

\item{dist.method}{distance method when using `pcoa`, default to 'manhattan'}

\item{point.size}{point size for pcoa}
}
\value{
ggplot bar diagram or heatmap of gene usage
}
\description{
This function will allow for the visualizing the distribution 
of the any VDJ and C gene of the TCR or BCR using heatmap or
bar chart. This function requires assumes two chains were used in 
defining clonotype, if not, it will default to the only chain 
present regardless of the chain parameter.
}
\examples{
#Making combined contig data
combined <- combineTCR(contig_list, 
                        samples = c("P17B", "P17L", "P18B", "P18L", 
                                    "P19B","P19L", "P20B", "P20L"))

vizGenes(combined, gene = "V", chain = "TRB", plot = "bar", scale = TRUE)

}
